Работа с Коллекциями и тестирование JUnit

CzechJD opened the pull request.

CzechJD committed:
Работа с Коллекциями и тестирование JUnit

psorokin02 commented on src/main/java/com/example/collections/Book.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Objects;
+
+public class Book {
+    private String title;
```
Давай все внутренние поля, которые не изменяются будем делать final - это best practice

psorokin02 commented on src/main/java/com/example/collections/User.java:
```diff
@@ -0,0 +1,43 @@
+
+public class User {
+    private String name;
+    private int age;
```
Аналогично сделать final

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.HashMap;
+
+public class LibraryService {
+    private HashMap<Long, Book> books;
```
final

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+    private HashMap<Long, User> users;
+    private HashMap<Long, ArrayList<Long>> userBorrowedBooks = new HashMap<>();
+
+    public LibraryService() {
```
Инициализация должна быть извне, а не внутри конструктора. 
Представь ты пишешь приложение, которые управляет разными библиотеками в городе. Ты хочешь создать одну библиотку с одним набором пользователей и книгами, а вторую - с другим набором пользователей и книг. С твоим подходом у тебя библиотека всегда создается с одним набором людей и книг
То как ты написал - связывает данные с реализацией. Это плохо
Нужно чтобы список книг приходил в конструкторе и список пользователей

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+
+    }
+
+    public static void main(String[] args) {
```
Main лучше отдельно в класс Main вынести

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+
+    public ArrayList<Book> getAllAvailableBooks() {
+        ArrayList<Book> availableBook = new ArrayList<>();
+        for (Book book : books.values()) {
```
Метод можно сделать быстрее, если дополнительно хранить
Map<long, Boolean> - ключ - id книги, значение - взята ли какая-то книга
Тогда получится проход в один цикл

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+        }
+        ArrayList<Book> listBook = new ArrayList<>();
+
+        if (userBorrowedBooks.containsKey(userId)) {
```
давай просто вернем тут сразу список userBorrowedBooks.get(userId); Без лишних операций.
Тогда нужно сразу хранить List<Book>, а не список Id-шников

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+        return listBook;
+    }
+
+    public boolean takeBook(Long userId, Long bookId) {
```
Хорош 👍

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+        } else if (!books.containsKey(bookId)) {
+            throw new IllegalArgumentException("Книга с ID: " + bookId + ", не найдена");
+        }
+        if (userBorrowedBooks.containsKey(userId)) {
```
Смотри у тебя получается кейс, такой 
if (condition) {
МНОГО КОДА
}
return false

В таких случаях лучше сделать инверсию
if (!condition) {
   return false;
}
МНОГО КОДА ТУТ

Выглядеть будет более читаемо

psorokin02 commented on src/main/java/com/example/collections/LibraryService.java:
```diff
@@ -0,0 +1,109 @@
+            if (userBooks.contains(bookId)) {
+                userBooks.remove(bookId);
+                if (userBooks.isEmpty()) {
+                    userBorrowedBooks.remove(userId);
```
А зачем удалять из Map? пусть пустой список будет, все ок же

psorokin02 commented on src/test/java/test/collections/testCollections.java:
```diff
@@ -0,0 +1,54 @@
+public class testCollections {
+    @Test
+    public void getUserBooks() {
+        LibraryService libraryService = new LibraryService();
```
Нужно инициализировать библиотеку тут, а не в конструкторе. 
И тогда тесты будут выглядеть понятнее
` libraryService.takeBook(user1.getId(), book2.getId());`
У тебя сейчас какой-то непонятный набор цифр взявшихся изниоткуда

CzechJD committed:
Merge branch 'master' into collections

# Conflicts:
#	pom.xml

CzechJD committed:
1. Классы стали record, что позволяет сократить код, не потеряв функциональности
2. Добавлена структура данных HashMap<Long, Boolean>, чтобы определить, свободна книга или нет.
3. Исправлены методы:
- getAllBooks()
- getUserBooks()
- takeBook()
- returnBook()
в связи с добавлением новой структуры данных
4. Все методы исправлены согласно комментариям
5. Теперь пользователи и книги добавляются в классе Main отдельно к каждому объекту, а не создаются вместе с объектом, как было раньше.
6. Добавлен методы addBook() и addUser() для добавления книг и пользователей к определённому объекту
7. Исправлены тесты: теперь для каждого теста создаётся отдельный объект Библиотеки.
8. Добавлены геттеры к классам: User, Book, LibraryService

CzechJD committed:
1. Классы стали record, что позволяет сократить код, не потеряв функциональности
2. Добавлена структура данных HashMap<Long, Boolean>, чтобы определить, свободна книга или нет.
3. Исправлены методы:
- getAllBooks()
- getUserBooks()
- takeBook()
- returnBook()
в связи с добавлением новой структуры данных
4. Все методы исправлены согласно комментариям
5. Теперь пользователи и книги добавляются в классе Main отдельно к каждому объекту, а не создаются вместе с объектом, как было раньше.
6. Добавлен методы addBook() и addUser() для добавления книг и пользователей к определённому объекту
7. Исправлены тесты: теперь для каждого теста создаётся отдельный объект Библиотеки.
8. Добавлены геттеры к классам: User, Book, LibraryService
